Solutions/Marain.ContentManagement.Api.Services.AspNetCore/Microsoft/Extensions/DependencyInjection/ContentManagementServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,102 @@
+        /// </param>
+        /// <param name="configureHost">Optional callback for additional host configuration.</param>
+        /// <returns>The service collection, to enable chaining.</returns>
+        public static IServiceCollection AddTenantedContentManagementApi(
+            this IServiceCollection services,
+            IConfiguration rootTenantDefaultConfiguration,
+            Action<IOpenApiHostConfiguration> configureHost = null)
+        {
+            Type contentServiceType = typeof(ContentService);
+
+            // If any of the OpenApi services are already installed, assume we've already completed installation and return.
+            if (services.Any(services => contentServiceType.IsAssignableFrom(services.ImplementationType)))
+            {
+                return services;
+            }
+
+            services.AddLogging();
+
+            services.AddContentManagementContent();
+
+            services.AddTenantCloudBlobContainerFactory(rootTenantDefaultConfiguration);
+            services.AddTenantProviderBlobStore();
+            services.AddTenantedAzureCosmosContentStore(rootTenantDefaultConfiguration);
+
+            services.AddContentManagementOpenApiServices();
+            services.AddContentManagementHalDocumentMappers();
+
+            services.AddOpenApiHttpRequestHosting<SimpleOpenApiContext>(config =>
+            {
+                config.AddContentManagementServiceDefinition();
+
+                configureHost?.Invoke(config);
+
+                config.MapTenancyExceptions();
+                config.MapContentManagementExceptions();
+            });
+
+            return services;
+        }
+
+        private static IServiceCollection AddContentManagementOpenApiServices(this IServiceCollection services)
+        {
+            services.AddSingleton<ContentService>();
+            services.AddSingleton<IOpenApiService, ContentService>(s => s.GetRequiredService<ContentService>());
+            return services;
+        }
+
+        private static IServiceCollection AddContentManagementHalDocumentMappers(this IServiceCollection services)
+        {
+            services.AddHalDocumentMapper<Content, IOpenApiContext, ContentMapper>();
+            return services;
+        }
+
+        private static void AddContentManagementServiceDefinition(this IOpenApiHostConfiguration config)
+        {
+            Type contentServiceType = typeof(ContentService);
+            using Stream apiYamlStream = contentServiceType.Assembly.GetManifestResourceStream($"{contentServiceType.Namespace}.ContentManagementServices.yaml");
+            var reader = new OpenApiStreamReader();
+            OpenApiDocument apiYamlDoc = reader.Read(apiYamlStream, out OpenApiDiagnostic diagnostic);
+            config.Documents.Add(apiYamlDoc);
+        }
+
+        private static void MapTenancyExceptions(this IOpenApiHostConfiguration config)
+        {
+            config.Exceptions.Map<TenantNotFoundException>(404);
```

Comment:
In this case, tenant is being supplied in the URL path, e.g. `/{tenantId}/marain/content/{slug}`. As a result, I believe it's correct to return a 404 if the tenant Id is incorrect.

Re: the bigger question on exception mapping - I've created an issue in the Menes repo for further discussion - https://github.com/menes-dotnet/Menes/issues/12.

Reply:
> In this case, tenant is being supplied in the URL path

I would argue that this is extremely non-discoverable. How is someone looking at this code meant to determine that the effect is as you describe? And even if they successfully determine that this is the effect, how are they meant to assess whether it was the intent?

This might be OK if the exception in question were specific to CMS. In cases where the exception is local to a particular app, there are situations in which a mapping-driven approach might make sense. But this `TenantNotFoundException` is defined by a different library which doesn't appear to document the places you can except the exception to occur. (As far as I can tell they all emerge from `ITenantProvider` but the XML docs for that don't currently tell you this. I've added https://github.com/corvus-dotnet/Corvus.Tenancy/issues/38 to track that.)

Tenancy is an infrastructure element, so it doesn't seem reasonable to presume that this exception could only possibly ever emerge from one particular endpoint handler. Even if today that exception only happens to emerge from there, it's plausible that we might add in our generic claims-based security at some point, and there are scenarios in which that will want to look up the tenant. If something is misconfigured, this could result in a `TenantNotFoundException` that should cause a 500 error, not a 404.

If the intent is that this particular exception handling applies only to a particular endpoint, I believe that right way is to `catch` is in the handler for that endpoint and return a 404 explicitly. Or, if you prefer, catch it in that handler, and throw a different CMS-specific exception and add a mapping for that instead.

Reply:
As we discussed earlier, we will put this discussion on hold while we discuss Menes global error handling over here: https://github.com/menes-dotnet/Menes/issues/26

Once that's resolved we can return to this in a separate issue.